auth0/src/main/java/com/auth0/android/authentication/JwkProvider.java

```diff
@@ -0,0 +1,184 @@
+package com.auth0.android.authentication;
+
+import android.net.Uri;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.VisibleForTesting;
+import android.support.annotation.WorkerThread;
+import android.util.Base64;
+import android.util.Log;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonSyntaxException;
+import com.google.gson.annotations.SerializedName;
+import com.google.gson.reflect.TypeToken;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.lang.reflect.Type;
+import java.math.BigInteger;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.Charset;
+import java.security.InvalidKeyException;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PublicKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.RSAPublicKeySpec;
+import java.util.ArrayList;
+import java.util.List;
+
+class JwkProvider implements KeyProvider {
+
+    private static final String TAG = JwkProvider.class.getSimpleName();
+    private final URL url;
+    private List<Jwk> jwks;
+
+    JwkProvider(@NonNull String domain) {
+        this(createUrl(domain));
+    }
+
+    @SuppressWarnings("WeakerAccess")
+    @VisibleForTesting
+    JwkProvider(@NonNull URL url) {
+        this.url = url;
+    }
+
+    @VisibleForTesting
+    URL getURL() {
+        return this.url;
+    }
+
+    private static URL createUrl(String domain) {
+        String safeUrl = domain.startsWith("http") ? domain : "https://" + domain;
```

Comment:
This would fail if the domain name starts with `http`:

https://http2.github.io/